subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/resolver/ConcurrentGroupingQueueTest.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.resolver
+
+import org.hamcrest.CoreMatchers.equalTo
+import org.hamcrest.MatcherAssert.assertThat
+
+import org.junit.Test
+
+import kotlin.concurrent.thread
+
+
+class ConcurrentGroupingQueueTest {
+
+    @Test
+    fun `groups similar requests ordered by most recent`() {
+
+        val requestsInGroup1 = (0..1).map { Request(1, it) }
+        val requestsInGroup2 = (0..1).map { Request(2, it) }
+
+        val subject = ConcurrentGroupingQueue<Request> { group == it.group }
+        for ((r1, r2) in requestsInGroup1 zip requestsInGroup2) {
+            // interleave the requests
+            subject.push(r1)
+            subject.push(r2)
+        }
+
+        assertThat(
+            subject.nextGroup(),
+            equalTo(requestsInGroup2.reversed())
+        )
+
+        assertThat(
+            subject.nextGroup(),
+            equalTo(requestsInGroup1.reversed())
+        )
+    }
+
+    @Test
+    fun `#nextGroup honours timeout`() {
+
+        val subject = ConcurrentGroupingQueue<Int> { false }
+
+        thread {
+            Thread.sleep(150)
```

Comment:
Instead of using a hardcoded sleep, you might want to consider using something like this library with a waiter.

https://github.com/jhalterman/concurrentunit

Reply:
Thanks for the link, would you have a suggestion on how to improve this test to take advantage of that lib?

Reply:
I'd reccomend doing something like this.

```kotlin
@Test
fun `#nextGroup honours push that happens in a different thread`() {

    val subject = ConcurrentGroupingQueue<Int> { false }
    val waiter = Waiter()

    thread {
        waiter.await()
        subject.push(42)
    }

    assertThat(
        subject.nextGroup(50),
        equalTo(emptyList())
    )
    waiter.resume()

    assertThat(
        subject.nextGroup(200),
        equalTo(listOf(42))
    )
}
```

Reply:
You may just be able to do this with a lock/semaphore though.

Reply:
+1 to finding a way to remove the hardcoded sleep

Reply:
I solved it with a `CountDownLatch` for simplicity.